test/test_controller.py

```diff
@@ -1282,3 +1283,198 @@ def test_collect_binlogs_to_purge():
     assert binlogs_to_purge == binlogs
     assert only_inapplicable_binlogs is False
     log.info.assert_any_call("Binlog %s has been replicated to all servers, purging", 3)
+
+
+@patch.object(RestoreCoordinator, "MAX_BASEBACKUP_ERRORS", 2)
+def test_restore_failed_basebackup(
+    default_backup_site,
+    master_controller,
+    mysql_empty,
+    session_tmpdir,
+):
+    m_controller, master = master_controller
+    new_controller = None
+
+    mysql_empty.connect_options["password"] = master.connect_options["password"]
+
+    try:
+        data_generator = DataGenerator(
+            connect_info=master.connect_options,
+            make_temp_tables=False,
+        )
+        data_generator.start()
+
+        m_controller.switch_to_active_mode()
+        m_controller.start()
+
+        def streaming_binlogs():
+            assert m_controller.backup_streams
+            assert all(bs.active_phase == BackupStream.ActivePhase.binlog for bs in m_controller.backup_streams)
+
+        while_asserts(streaming_binlogs, timeout=10)
+
+        m_controller.mark_backup_requested(backup_reason=BackupStream.BackupReason.requested)
+        data_generator.stop()
+
+        def all_backups_completed(expected_num):
+            backups = m_controller.state["backups"]
+            assert len(backups) == expected_num
+            assert all(backup["completed_at"] is not None for backup in backups)
+
+        while_asserts(lambda: all_backups_completed(expected_num=2), timeout=30)
+
+        m_controller.stop()
+
+        # get the most recent backup (corrupted)
+        corrupted_backup = sort_completed_backups(m_controller.state["backups"])[1]
```

Comment:
this is not an actually corrupted backup, is it? it's the side effect monkeypatching down there that makes it corrupted.